02-peer-protocol.md

```diff
@@ -560,6 +1084,346 @@ ones it has sent so it can use them should they be requested by
+  - if is the *accepter*:
+    - MUST fail the negotiation if:
+      - no funding output was received
+      - the value of the funding output is not equal to the sum of
+        `open_channel2`.`funding_satoshis` and `accept_channel2`.
+        `funding_satoshis`
+      - the value of the funding output is less than the `dust_limit`
+  - if this is an RBF attempt:
+    - MUST fail the negotiation if:
+      - the transaction's total fees is less than the last
+        successfully negotiated transaction's fees
+	  - the transaction does not share at least one input with
+	    each previous funding transaction
+  - if it has sent `require_confirmed_inputs` in `open_channel2` or `accept_channel2`:
+    - MUST fail the negotiation if:
+      - one of the inputs added by the other peer is unconfirmed
+
+### The `commitment_signed` Message
+
+This message is exchanged by both peers. It contains the signatures for
+the first commitment transaction.
+
+Rationale and Requirements are the same as listed below,
+for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.
+
+#### Requirements
+
+The sending node:
+  - MUST send zero HTLCs.
+  - MUST remember the details of this funding transaction.
+
+The receiving node:
+  - if the message has one or more HTLCs:
+    - MUST fail the negotiation
+  - if it has not already transmitted its `commitment_signed`:
+    - MUST send `commitment_signed`
+  - Otherwise:
+    - MUST send `tx_signatures` if it should sign first, as specified
+      in the [`tx_signatures` requirements](#the-tx_signatures-message)
+
+#### Rationale
+
+The first commitment transaction has no HTLCs.
+
+Once peers are ready to exchange commitment signatures, they must remember
+the details of the funding transaction to allow resuming the signatures
+exchange if a disconnection happens.
+
+### Sharing funding signatures: `tx_signatures`
+
+After a valid `commitment_signed` has been received
+from the peer and a `commitment_signed` has been sent, a peer:
+  - MUST transmit `tx_signatures` with their signatures for the funding
+    transaction, following the order specified in the
+    [`tx_signatures` requirements](#the-tx_signatures-message)
+
+#### Requirements
+
+The sending node:
+  - MUST verify it has received a valid commitment signature from its peer
+  - MUST remember the details of this funding transaction
+  - if it has NOT received a valid `commitment_signed` message:
+    - MUST NOT send a `tx_signatures` message
+
+The receiving node:
+  - if has already sent or received a `channel_ready` message for this
+    channel:
+    - MUST ignore this message
+  - if the `witness` weight lowers the effective `feerate`
+    below the *opener*'s feerate for the funding transaction and the effective
+	`feerate` is determined by the receiving node to be insufficient for
+	getting the transaction confirmed in a timely manner:
+    - SHOULD broadcast their commitment transaction, closing the channel
```

Comment:
As discussed on the other comment thread, I think we should remove the recommendation to broadcast the commitment, since that may actually punish ourselves more than the fee-underpaying peer.  And broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel.  It seems a little silly to force close a working channel out of spite.

The existing recommendation to productively double-spend if the fee is underpaid addresses the potential griefing vector.  If the funding tx confirms, then either no griefing happened in practice or it ended before we could productively use those funds anyway.  No harm, no foul, so let's not make things worse by force closing?

```suggestion
```

Also, (nit) the indentation is off in the recent changes here.

Reply:
> broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel. It seems a little silly to force close a working channel out of spite.

That's not entirely correct: I'd expect the commitment transaction to sit in the mempool along with the unmined opening transaction. In the case that the channel does open, it will be closed by the commitment transaction. I don't see any downsides to broadcasting a close in the case that you're attempting to re-use inputs in the funding elsewhere. You've already so much as considered this channel dead on arrival, there's no harm in publishing a transaction you're not anticipating will be mined anyway.

Also note, the force close is recommended but not mandatory behavior. There's a real benefit to the peer knowing that if they underpay the channel likely won't open.

Reply:
I feel like we're talking past each other.  I'll try to respond directly to each of your comments below.  Please let me know where I'm incorrect or missing something.

> > broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel. It seems a little silly to force close a working channel out of spite.
> 
> That's not entirely correct: I'd expect the commitment transaction to sit in the mempool along with the unmined opening transaction. In the case that the channel does open, it will be closed by the commitment transaction.

I think it's exactly correct.  As you've just said, the commitment tx can only confirm if the channel opens.  If the channel never opens, and therefore the commitment tx never confirms, does the fact that you broadcasted commitment tx really matter?

And if the funding tx does confirm, any griefing is over, so there's little benefit to force closing at that point.

> I don't see any downsides to broadcasting a close in the case that you're attempting to re-use inputs in the funding elsewhere. You've already so much as considered this channel dead on arrival, there's no harm in publishing a transaction you're not anticipating will be mined anyway.

Here's 3 downsides to broadcasting the commitment tx:

1. Loss of funds -- If you're the opener and the commitment tx confirms, you probably pay more fees (funding fees + commitment fees) than if you had double spent (funding fees only).  This is especially true if your node does CPFP on the anchor output of the commitment, which I suspect most implementations currently do by default.
2. Loss of opportunity -- If the funding tx confirms, you could have actually used that channel.  But, by broadcasting the commitment tx, you've preemptively force-closed the channel instead.
3. More complexity -- Teaching your node to do 2 things (broadcast commitment + double spend) is more complex than teaching it to do 1 thing (double spend).

> Also note, the force close is recommended but not mandatory behavior.

Yes, but the recommendation in the spec is important.  We should not expect future implementers to find this PR and read our current discussion about tradeoffs before implementing.

> There's a real benefit to the peer knowing that if they underpay the channel likely won't open.

Yes, and I think the recommendation to double spend achieves this well enough on its own.  The peer knows that if they underpay, their counterparty will double spend at their earliest convenience, therefore the channel likely won't open.

Reply:
> I feel like we're talking past each other. 

From where I'm sitting, it appears we fundamentally disagree about what the proper behavior should be. 

Ultimately I think it's crucial to remember that underpaying fees is a bug. An implementation can avoid the undesirable outcomes of this situation by paying the proper amount for any witness data they add to a transaction.

Having a node force close for underpayment makes it much easier to spot + fix bugs in implementations, which is ultimately the goal imo and hence the spec recommendation.

Reply:
I disagree about this being a practical griefing vector. It's costly for a peer to execute: they would have had to include inputs that they're paying for in order to trigger this, and they would have had to paid for everything *except* the witness data at the agreed fee rate for the transaction. Why would they do all that just to 'maliciously' underpay on witness data at the last moment? Seems like behavior that the attempt of should be discouraged.

Also, the risk of your peer sending you a channel error or asking you to close the channel exist in other parts of the protocol, this isn't a new 'griefing' vector by any means.

Reply:
> I disagree about this being a practical griefing vector. It's costly for a peer to execute: they would have had to include inputs that they're paying for in order to trigger this, and they would have had to paid for everything except the witness data at the agreed fee rate for the transaction.

That's a good point, it's not completely free for the attacker either, they also have to pay on-chain fees.
Since those requirements are a `SHOULD`, that works for me then, I don't mind having it and letting implementations decide whether they want to broadcast their commit or not!

Reply:
> And I agree that the severity of this griefing vector is low relative to the simpler and more effective griefing vectors in the interactive-tx protocol -- e.g., the funding tx can be pinned in the mempool, requiring the aggrieved party to burn a ridiculous amount of fees to double-spend the funding tx.

Hmm... I think I just convinced myself there's a scenario where broadcasting the commitment tx makes *more* sense than trying to double spend.

If the griefer inflates their witness such that the funding tx will never confirm and also pins the funding tx with minimum fee-rate descendants, the victim's fee cost to double spend the funding tx can become ridiculous.  But if the victim does CPFP on the funding tx instead, they can get it to confirm with relatively low fees.  The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

Perhaps we should mention this scenario and mitigation in the spec?

Reply:
CC: @ariard

Reply:
Pinning, it's always pinning... :sweat_smile: 

> The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

True, but it wouldn't always work though: if the funding tx is evicted from the mempool, you won't be able to CPFP it with the commitment + anchor (even package relay won't work, because it only handles two generations of txs, not three).

But that's ok! Because if the funding tx is evicted from the mempool, then it becomes free to double-spend (the pinning is gone). So your strategy in that case should be to keep trying both options in parallel, and one of them should eventually work.

Reply:
> The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

Long(er) chains of transactions are less likely to propagate in general properly, and future opt-in like V3 policies may disallow this entirely, so all else equal stick to shorter CPFPs.

Reply:
I'm afraid there's no good defense against pinning the funding tx.  The griefer can make RBF prohibitively expensive *and* also block CPFP by saturating the package size limit.

For commitment txs, we have the CSV 1 trick and anchors with carve out.  But we don't have a great way to enforce CSV 1 on funding tx outputs -- and probably we don't want to.

For commitment txs, pinning can cause loss of funds due to HTLC timeout.  But for funding txs the problem is griefing (fund lockup) rather than loss of funds, and to carry out the griefing the attacker must be willing to pay significant fees eventually.  Perhaps we can just accept that if the attacker is willing to burn enough fees, they can lock up our funds for a long time.

We actually [previously discussed](https://github.com/niftynei/lightning-rfc/pull/12#discussion_r1102962560) this class of griefing vectors against the interactive-tx protocol.  Back then, we agreed that "just live with it" may be the best course for now.

--------------------------------

My suggestion for the spec is to recommend both double spending and CPFP as methods to unlock funds or get the funding tx to confirm, but to note that neither is guaranteed to work if the peer is actively trying to grief us.

Reply:
Revisiting this...

For simplicity, I think we can leave discussion of pinning out of the spec for now.  Since we have no great defense against it, there's not much to discuss.

But I still think the current suggested handling of witness inflation is bad -- by broadcasting the commitment tx, we force our future double spend attempts to pay ~2x the fee (RBF rule #3).  So if the peer is just buggy (not malicious), this is an expensive way to recover our funds.

Reply:
I agree, those are subtle issues that aren't obvious to a spec reader today. How would you address that? By adding more details in the rationale section? Can you propose a patch that you think makes it better?

Reply:
I suggested a change on the first comment in this thread.  I think removing that single line is sufficient.

If you think there's more we should include, I can attempt a larger patch.

Reply:
> My suggestion for the spec is to recommend both double spending and CPFP as methods to unlock funds or get the funding tx to confirm, but to note that neither is guaranteed to work if the peer is actively trying to grief us.

I think that made the most sense instead of simply removing `SHOULD broadcast their commitment transaction`, but that requires explaining that it depends on the situation. You don't think we should suggest both options anymore?

Reply:
I think there's two potential situations -- buggy implementation and malicious liquidity griefing.

#### Buggy implementation

Double spending opportunistically is the best course.  Broadcasting the commitment transaction only increases our costs of double spending.

#### Liquidity griefing

We have no foolproof defense against this currently, and any RBF or CPFP recommendation can be defeated by pinning the inflated funding transaction.

Unless we are going to recommend one of the solutions [discussed](https://delvingbitcoin.org/t/liquidity-griefing-in-multi-party-transaction-protocols/264#tldr-1) on Discourse (i.e. initiator signs first, require p2wpkh inputs, or CSV 1 on outputs), our best course of action is probably to just unlock the UTXOs and hope our double spends somehow get mined, perhaps because the pinned tx is evicted from mempools.

Reply:
That makes sense to me, on the eclair side I'm planning to just unlock the utxos in that case. I'm fine with removing the `SHOULD` then if we convince @niftynei !